ed2nav: Writes a temp file before replacing the old data

BaptisteB-A opened the pull request:
Now serializes the data in a temporary file, and replaces the
previous `data.nav.lz4` file. In case of error, the temporary file is
deleted instead of the old data file as it used to be, so the old data
file doesn't get corrupted.

JIRA : https://jira.kisio.org/browse/NAVP-1301

BaptisteB-A committed:
Writes a temp file before replacing the old data

Now serializes the data in a temporary file, and replaces the
previous `data.nav.lz4` file. In case of error, the temporary file is
deleted instead of the old data file as it used to be, so the old data
file doesn't get corrupted.

buildbotctp commented:
Can one of the admins verify this patch?

kinnou02 commented:
Ok to test

kinnou02 commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +165,33 @@ struct FindAdminWithCities {
+    try {
+        data.save(temp_output);
+        try {
+            remove(output_filename.c_str());
```
If this is `remove` from `stdio` you need to check the return value, same things for `rename`.

kinnou02 commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +165,33 @@ struct FindAdminWithCities {
+            remove(output_filename.c_str());
+            rename(temp_output.c_str(), output_filename.c_str());
+            LOG4CPLUS_INFO(logger, "Data saved");
+        } catch (const navitia::exception& e) {
```
nothing in this block will throw a `navitia::exception`

kinnou02 commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +165,33 @@ struct FindAdminWithCities {
+    try {
+        data.save(temp_output);
+        try {
+            remove(output_filename.c_str());
```
You should probably rename it first.

BaptisteB-A replied:
I see what you mean but if I rename `temp_output` file name to `output_filename`, and then remove the `output_filename` file, wouldn't there be a problem ?

kinnou02 replied:
What I meant was to rename the current `data.nav.lz4` to something like `data.nav.lz4.bak`, then rename `data.nav.lz4.temp`  to `data.nav.lz4`, this way if the renaming fail for whatever reason we still have a valid dataset to recover.

BaptisteB-A committed:
Replaced try/catch by if/else to check `rename` & `remove` returns

kinnou02 commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +165,31 @@ struct FindAdminWithCities {
+    try {
+        data.save(temp_output);
+        if (remove(output_filename.c_str()) == 0) {
+            if (rename(temp_output.c_str(), output_filename.c_str()) == 0)
```
Always put brackets, it doesn't cost anything and reduce the risk of bug later on.

kinnou02 commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +165,31 @@ struct FindAdminWithCities {
+        if (remove(temp_output.c_str()) == 0)
+            LOG4CPLUS_INFO(logger, "Temp data removed because it was corrupted");
+        else
+            LOG4CPLUS_ERROR(logger, "Error deleting temp file: No such file or directory");
```
you log an error that may or may not be the one that occured, it will mislead people if the error is caused by something else like a filesystem in readonly. You should use `strerror` to get the correct error message

BaptisteB-A committed:
More precise errors, keeps a backup of the previous data, ...

Uses `strerror`s instead of a custom error in log
Adds `.bak` to the previous data file in case we need it to recover
New test : `writing_test_ed2nav_should_not_throw`

pbougue commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,42 @@ struct FindAdminWithCities {
+bool write_data_to_file(const std::string& output_filename, navitia::type::Data& data) {
+    auto logger = log4cplus::Logger::getInstance("log");
+    std::string temp_output_filename = output_filename + ".temp";
+    std::string backup_output_filename = output_filename;
```
Shouldn't it be 
```suggestion
    std::string backup_output_filename = output_filename + ".bak";
```
?
(then used below)

BaptisteB-A replied:
I didn't make the names very clear but I wanted to store the old output_filename in back_output_filename before adding ".bak" to the old output_filename. The new data file name with temp_output_filename is then renamed into back_output_filename, or, if you prefer, the original output_filename. Hum, sorry I'm not sure if that really is clearer..

pbougue commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,42 @@ struct FindAdminWithCities {
+        }
+        LOG4CPLUS_INFO(logger, "Temp file removed");
+        if (rename(output_filename.c_str(), (backup_output_filename).c_str()) != 0) {
+            LOG4CPLUS_INFO(logger, "Unable to rename old data file: " << std::strerror(errno));
```
Why info more than error?

BaptisteB-A replied:
Fixed

pbougue commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,42 @@ struct FindAdminWithCities {
     }
 };
 
+bool write_data_to_file(const std::string& output_filename, navitia::type::Data& data) {
```
I'd like more a `void` method raising if there is an error, than a `bool` way of managing errors.

BaptisteB-A replied:
Why not a `bool` ?

pbougue replied:
As discussed aside, no need to change here.
To trace it a bit : the conventional way of dealing with errors in C++ (and navitia) is more throwing exception.

pbougue commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,42 @@ struct FindAdminWithCities {
+        LOG4CPLUS_INFO(logger, "Old data file successfully renamed to " << output_filename << ".bak");
+    }
+    try {
+        data.save(temp_output_filename);
```
Why not writing it into output_filename directly (as it has been mv)?

pbougue commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,42 @@ struct FindAdminWithCities {
+            return false;
+        }
+        LOG4CPLUS_INFO(logger, "Temp file removed");
+        if (rename(output_filename.c_str(), (backup_output_filename).c_str()) != 0) {
```
It's the opposite, no?

BaptisteB-A replied:
No, we want the old file to be renamed after back_output_filename.

pbougue commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,42 @@ struct FindAdminWithCities {
+        LOG4CPLUS_ERROR(logger, e.what());
+        if (remove(temp_output_filename.c_str()) != 0) {
+            LOG4CPLUS_ERROR(logger, "Unable to remove temp data file: " << std::strerror(errno));
+            return false;
```
Here you still have to mv backup file to it's original name, it's too early to return I think.

BaptisteB-A committed:
More precise writing binary test (with comments)

BaptisteB-A committed:
Replaced 'include type/data.h' by a forward declaration

BaptisteB-A committed:
Pre-commit run on ed2nav.h

benoit-bst commented:
Jenkins, test this please !

pbougue commented on source/ed/tests/ed2nav_test.cpp:
```diff
@@ -50,3 +53,26 @@ BOOST_AUTO_TEST_CASE(should_throw_on_bad_connection_string) {
     BOOST_CHECK_THROW(ed::ed2nav(2, argv), std::exception);
 }
+
+BOOST_AUTO_TEST_CASE(writing_test_ed2nav_should_not_throw) {
```
Can you add failing tests?
(like rename impossible, file not generated, etc.)

BaptisteB-A replied:
Added a test on file save

nonifier replied:
Unfortunately, making a call to `rename` to fail (in a decent way), would mean to mock the free function which is very complicated without a proper mocking  framework :cry:

BaptisteB-A committed:
Added test on data saving failure

BaptisteB-A commented:
Jenkins, test this please !

xlqian commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,41 @@ struct FindAdminWithCities {
 };
 
+bool rename_old_file(const std::string& output_filename, std::string& backup_output_filename) {
+    auto logger = log4cplus::Logger::getInstance("log");
```
you can give a more specific name for the logger 
`auto logger = log4cplus::Logger::getInstance("ed2nav::rename_old_file");`

xlqian commented on source/ed/ed2nav.h:
```diff
@@ -27,11 +27,39 @@ IRC #navitia on freenode
 namespace ed {
 
+template <class T = navitia::type::Data>
+bool try_save_file(std::string& filename, T& data) {
```
`const T& data` ?

xlqian commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,41 @@ struct FindAdminWithCities {
+}
+
+template <class T = navitia::type::Data>
+bool write_data_to_file(const std::string& output_filename, T& data) {
```
`const T& data` ?

xlqian commented on source/ed/ed2nav.h:
```diff
@@ -27,11 +27,39 @@ IRC #navitia on freenode
+bool rename_temp_file(std::string& temp_output_filename, std::string& output_filename);
+
+template <class T = navitia::type::Data>
+bool write_data_to_file(const std::string& output_filename, T& data);
```
`const T& data` ?

BaptisteB-A committed:
More specific names for loggers and added some `const`

kadhikari commented on source/ed/ed2nav.cpp:
```diff
@@ -165,6 +164,41 @@ struct FindAdminWithCities {
+    }
+    LOG4CPLUS_INFO(logger, "File: " << output_filename << " saved");
+    return true;
+}
```
Minor: Two functions above can be mutualized non?

BaptisteB-A replied:
Right, thanks !

BaptisteB-A committed:
rename_file function is now generic + edited writing test

AurelienLP commented:
Jenkins, test this please.

AurelienLP commented:
Jenkins, test this please.

mbounabi commented:
Jenkins, test this please.
